crates/sui-core/src/authority.rs

```diff
@@ -1590,100 +1592,173 @@ impl AuthorityState {
                 error: "dev-inspect is only supported on fullnodes".to_string(),
             });
         }
 
+        if transaction_kind.is_system_tx() {
+            return Err(SuiError::UnsupportedFeatureError {
+                error: "system transactions are not supported".to_string(),
+            });
+        }
+        let skip_checks = skip_checks.unwrap_or(true);
+        let reference_gas_price = epoch_store.reference_gas_price();
         let protocol_config = epoch_store.protocol_config();
-        transaction_kind.check_version_supported(protocol_config)?;
-        transaction_kind.validity_check(protocol_config)?;
-
         let max_tx_gas = protocol_config.max_tx_gas();
-        let reference_gas_price = epoch_store.reference_gas_price();
-        let gas_price = match gas_price {
-            None => reference_gas_price,
-            Some(gas) => {
-                if gas == 0 {
-                    reference_gas_price
-                } else {
-                    gas
-                }
-            }
+
+        let gas_price = gas_price.unwrap_or(reference_gas_price);
+        let gas_budget = gas_budget.unwrap_or(max_tx_gas);
+        let owner = gas_sponsor.unwrap_or(sender);
+        // Payment might be empty here, but it's fine we'll have to deal with it later after reading all the input objects.
+        let payment = gas_objects.unwrap_or_default();
+        let expiration = match epoch {
+            None => TransactionExpiration::None,
+            Some(epoch) => TransactionExpiration::Epoch(epoch),
         };
-        let gas_status =
-            SuiGasStatus::new(max_tx_gas, gas_price, reference_gas_price, protocol_config)?;
-
-        let gas_object_id = ObjectID::random();
-        // give the gas object 2x the max gas to have coin balance to play with during execution
-        let gas_object = Object::new_move(
-            MoveObject::new_gas_coin(SequenceNumber::new(), gas_object_id, max_tx_gas * 2),
-            Owner::AddressOwner(sender),
-            TransactionDigest::genesis_marker(),
-        );
+        let transaction = TransactionData::V1(TransactionDataV1 {
+            kind: transaction_kind.clone(),
+            sender,
+            gas_data: GasData {
+                payment,
+                owner,
+                price: gas_price,
+                budget: gas_budget,
+            },
+            expiration,
+        });
 
-        let input_object_kinds = transaction_kind.input_objects()?;
-        let receiving_object_refs = transaction_kind.receiving_objects();
-        let (input_objects, receiving_objects) = self
+        transaction.check_version_supported(protocol_config)?;
+        transaction.validity_check_no_gas_check(protocol_config)?;
+
+        let input_object_kinds = transaction.input_objects()?;
+        let receiving_object_refs = transaction.receiving_objects();
+
+        if !skip_checks {
+            sui_transaction_checks::deny::check_transaction_for_signing(
```

Comment:
IIUC, I think we should perform this particular check always -- because it's really more a system protection method.